tools/ci-build/publisher/src/package.rs

```diff
@@ -255,38 +340,60 @@ fn read_package(path: &Path, manifest_bytes: &[u8]) -> Result<Option<Package>> {
 /// Validates that all of the publishable crates use consistent version numbers
 /// across all of their local dependencies.
 fn validate_packages(packages: &[Package]) -> Result<()> {
-    let mut versions: BTreeMap<String, Version> = BTreeMap::new();
-    let track_version = &mut |handle: &PackageHandle| -> Result<(), Error> {
-        if let Some(version) = versions.get(&handle.name) {
+    let mut versions: BTreeMap<String, LenientVersion> = BTreeMap::new();
+    let mut version_requirements: BTreeMap<String, LenientVersion> = BTreeMap::new();
+    let track_version = &mut |handle: &PackageHandle,
+                              map: &mut BTreeMap<String, LenientVersion>|
+     -> Option<(String, LenientVersion, LenientVersion)> {
```

Comment:
Consider returning a result instead of an option since you're mapping it into an error later anyway.

Reply:
I think I'm looking for a nicer way myself here, but if this returns a result, the implementation needs to know who calls `track_version` because we're mapping a returned tuple to different error enum variants; if it's called with a map `versions` then it ends up `Error::MultipleVersions` but if it's called with a map `version_requirements`, then it ends up with `Error::MultipleVersionRequirements`. Being inside `track_version`, we don't know which error variant should be returned.

Reply:
Maybe I can pass in a closure that returns `Error` to `track_version`, and each call site gets to specify what error variant it wants the function to return.

Reply:
Maybe I can pass in to `track_version` a closure that takes a tuple and returns `Error`, and each call site gets to specify what error variant it wants the function to return.

Reply:
Addressed in 1597129.